pkg/grub/Dockerfile

```diff
@@ -19,32 +19,39 @@ RUN ln -s python3 /usr/bin/python
 
 # list of grub modules that are portable between x86_64 and aarch64
 ENV GRUB_MODULES_PORT="part_gpt fat ext2 iso9660 squash4 gzio linux acpi normal cpio crypto disk boot crc64 gpt \
-search_disk_uuid search_part_label search_label verify xzio xfs video gfxterm efi_gop gptprio chain probe reboot regexp smbios    \
+search_disk_uuid search_part_label search_label verify xzio xfs video gfxterm serial gptprio chain probe reboot regexp smbios    \
 part_msdos cat echo test configfile loopback"
-ENV GRUB_MODULES_x86_64="multiboot multiboot2 efi_uga"
-ENV GRUB_MODULES_aarch64="xen_boot"
+ENV GRUB_MODULES_i386_pc="multiboot multiboot2 biosdisk"
+ENV GRUB_MODULES_x86_64="multiboot multiboot2 efi_uga efi_gop"
+ENV GRUB_MODULES_aarch64="xen_boot efi_gop"
 ENV GRUB_COMMIT=71f9e4ac44142af52c3fc1860436cf9e432bf764
 ENV GRUB_REPO=https://git.savannah.gnu.org/git/grub.git
 
 COPY patches/* /patches/
-RUN mkdir /grub-lib && \
-  set -e && \
-  git clone ${GRUB_REPO} && \
-  cd grub && \
-  git checkout -b grub-build ${GRUB_COMMIT} && \
-  for patch in /patches/*.patch; do \
-    echo "Applying $patch"; \
-    patch -p1 < "$patch"; \
-  done && \
+RUN mkdir /grub-lib && git clone ${GRUB_REPO}
+
+WORKDIR /grub
+RUN git config --add user.email a@b.c && git config user.name a && \
+    git checkout -b grub-build ${GRUB_COMMIT} && git am /patches/*
+
+ENV BUILD_GRUB="set -e && git clean -f -d -x && git reset --hard HEAD && \
   ./autogen.sh && \
-  ./configure --libdir=/grub-lib --with-platform=efi CFLAGS="-Os -Wno-unused-value" && \
-  make -j "$(getconf _NPROCESSORS_ONLN)" && \
-  make install
+  ./configure --libdir=/grub-lib --with-platform=\${GRUB_PLATFORM} CFLAGS='-Os -Wno-unused-value' && \
+  make -j $(getconf _NPROCESSORS_ONLN) && \
+  make install"
+
+ENV GRUB_PLATFORM=efi
+RUN eval $BUILD_GRUB
+RUN if [ "$(uname -m)" = x86_64 ]; then GRUB_PLATFORM="pc --disable-efiemu" && eval "$BUILD_GRUB" ; fi
```

Comment:
This strikes me as a bit odd, but maybe I didn't understand. Are we not building a grub that can run on both x86_64 and i386 (and arm64)? Why are we checking the build machine's architecture?

Reply:
It is impossible to build GRUB that way -- you can build it either as a UEFI payload or as a legacy BIOS payload. Of course, legacy BIOS payload doesn't apply to ARM -- hence the check.